### PR TITLE
LPS-118505 Wrong ellipsis button style on hover

### PR DESCRIPTION
**Pull request by dgarciasarai**
**Description**

After a conversation with @susanavr and @drakonux, the ellipsis button should have:

* Size: 32x32px 
* Style on hover: change color and background color

This occurs in the search container dropdown items in two display types: list and table.

**Fix**

* Wrong ellipsis button size and style on hover: https://issues.liferay.com/browse/LPS-118505

**Lexicon definition**

lexicon-list

**Before**

* Size: 40x40px
* No changes on hover

Screenshot 2020-08-06 at 11 33 23 Screenshot 2020-08-06 at 11 34 17

**After**

* Size: 32x32px
* Color and background color change on hover

Screenshot 2020-08-06 at 11 33 33 Screenshot 2020-08-06 at 11 34 03

### CODE REVIEW

**Comment by dgarciasarai** [comment]
Hey @jbalsas, I think you created the `DropdownActionsTag`, can you review this pull request? These changes are only an approach to what we want to achieve. Feel free to change it. Thanks!

**Comment by dgarciasarai** [comment]
ci:test:sf

**Comment by dgarciasarai** [comment]
ci:test:relevant

**Comment by carloslancha** [comment]
@dgarciasarai  according to [Clay Docs about Action Dropdowns](https://next.clayui.com/docs/components/drop-down.html#css)

```
<div class="dropdown dropdown-action">
	<button
		aria-expanded="false"
		aria-haspopup="true"
		class="component-action dropdown-toggle"
		id="dropdownBtnAction1"
		type="button"
	>
		<svg
			class="lexicon-icon lexicon-icon-ellipsis-v"
			focusable="false"
			role="presentation"
		>
			<use href="/images/icons/icons.svg#ellipsis-v"></use>
		</svg>
	</button>
	<ul aria-labelledby="dropdownBtnAction1" class="dropdown-menu">
		<li><a class="dropdown-item" href="#1">Download</a></li>
		<li><a class="dropdown-item" href="#1">Edit</a></li>
		...
	</ul>
</div>
```

- Dropdown container should have `dropdown dropdown-action` classes
- Trigger button should have `component-action dropdown-toggle` classes
- Button is not small
- Button is not outline

Could you please take a look?

Thx!

**Comment by dgarciasarai** [comment]
Hey @carloslancha, thanks for your feedback :) I updated the pull just changing the size and adding `component-action` CSS class.

> Dropdown container should have dropdown dropdown-action classes

That `dropdown-action` class is not present in any dropdown using `DropdownActionsTag`. I don't know where is the right place to change it so we can open an issue on Clay. What do you think?

> Trigger button should have component-action dropdown-toggle classes

Perfect, and for that I added `component-action` as a CSS class.

> Button is not small

I review the docs and the size it's 32x32:

Screenshot 2020-08-06 at 17 06 08

For that reason, I put `setSmall(true)` in the code. Normal size is 40x40 and that's not what we want. Here I have the actual ellipsis in the Content Dashboard with this commit:

Screenshot 2020-08-06 at 17 19 07

> Button is not outline

Thanks! I removed that part 👍 

===

If everything makes sense now, please, run the CI 😄 Thanks!

**Comment by dgarciasarai** [comment]
Also, If I'm not wrong, I saw that the ellipsis icon is not right. It seems 1 pixel lower than it should 🤔 

Screenshot 2020-08-06 at 17 01 13

cc: @pat270 @drakonux

**Comment by dgarciasarai** [comment]
ci:test:sf

**Comment by carloslancha** [comment]
@dgarciasarai 

- Once the dropdown has `dropdown-action` class the size of the trigger will be 32px, that's why in the markup example there's no `btn-sm` (so is not small).
- You can take a look to `modules/apps/frontend-taglib/frontend-taglib-clay/src/main/resources/META-INF/resources/DropdownMenu.js`, where we're rendering the dropdown.
- As you can see we're rendering a `<ClayDropDownWithItems` which accepts `className`, there's where we should pass `dropdown-action` in case the `DropdownMenu` is rendered from `DropdownActionsTag`
- The missalignment is caused by this span:
```
<span
	className={classNames({
		'inline-item': label,
		'inline-item-before': label,
	})}
>
	<ClayIcon symbol={icon} />
</span>
```
That in this case shouldn't be rendered (label is undefined)

**Comment by dgarciasarai** [comment]
> Once the dropdown has dropdown-action class the size of the trigger will be 32px, that's why in the markup example there's no btn-sm (so is not small).

Ok, I'll remove the small size from the code

> As you can see we're rendering a <ClayDropDownWithItems which accepts className, there's where we should pass dropdown-action in case the DropdownMenu is rendered from DropdownActionsTag

If I'm understand correct, we should pass the `className` attribute (which is not present at the moment) to `ClayDropDownWithItems` to add `dropdown-action` to the component. How can we check if we are rendering it from `DropdownActionsTag`?

> The misalignment is caused by this span

For this, maybe we can change to something like this? 🤔 

```js
trigger={
	<ClayButton className={cssClass} {...otherProps}>
		{icon &&
			(label ? (
				<span
					className={'inline-item inline-item-before'}
				>
					<ClayIcon symbol={icon} />
				</span>
			) : (
				<ClayIcon symbol={icon} />
			))}

		{label}
	</ClayButton>
}
```

**Comment by carloslancha** [comment]
> If I'm understand correct, we should pass the className attribute (which is not present at the moment) to ClayDropDownWithItems to add dropdown-action to the component. How can we check if we are rendering it from DropdownActionsTag? Of it's just like className="dropdown-action"?

I can think in some options here
- Add a `String type = "actions"` to the java class and receive it in `DropdownMenu.js`, then pass to `ClayDropDownWithItems` `className='dropdown-action'` checking that `type` param.
- Add a `String dropdownCssClass = "dropdown-action"` to the java class and receive it in `DropdownMenu.js`, then pass it to `ClayDropDownWithItems` `className={dropdownCssClass}`.
- Create a new React component `DropdownActions` that renders `DropdownMenu` passing it all the params plus `dropdownClassName`, receive that new param in `DropdownMenu` and pass it to `ClayDropDownWithItems` `className={dropdownCssClass}`.

In addition I'd like to point out that, to me, the `cssClass` attribute should be `triggerCssClass` since it's being applied to the trigger, not to the dropdown. That attribute existed before but was deprecated for 7.3 in favor of `cssClass`, don't know the reason for this decision (probably 'cause we don't want this to be configured via taglib, @jbalsas should have more info about it). Anyway it's not the moment to think about changing any tag API, just wanted to rise this up so we have a full understanding about it.

>I think that the span should only be rendered when there is a label

Exactly.

**Comment by dgarciasarai** [comment]
Hey @carloslancha @jbalsas, I just updated the pull request :) Can you review it? Thanks!

Screenshot 2020-08-10 at 16 24 56

**Comment by dgarciasarai** [comment]
ci.test:sf

**Comment by dgarciasarai** [comment]
ci:test:relevant

File: modules/apps/content-dashboard/content-dashboard-web/src/main/resources/META-INF/resources/view.jsp [hunk 1]
```diff
@@ -201,6 +201,7 @@ ContentDashboardAdminManagementToolbarDisplayContext contentDashboardAdminManage
 
 							<liferay-ui:search-container-column-text>
 								<clay:dropdown-actions
+									dropdownCssClass="dropdown-action"
```
Comment (carloslancha) [inline modules/apps/content-dashboard/content-dashboard-web/src/main/resources/META-INF/resources/view.jsp:204]: This looks like something that we shouldn't need to specify on each Tag use since according to clay markup all actions dropdowns should have that css class.

File: modules/apps/frontend-taglib/frontend-taglib-clay/src/main/java/com/liferay/frontend/taglib/clay/servlet/taglib/DropdownActionsTag.java [hunk 2]
```diff
@@ -25,6 +25,7 @@
 	public int doStartTag() throws JspException {
 		setAttributeNamespace(_ATTRIBUTE_NAMESPACE);
 
+		setCssClass("component-action");
```
Comment (carloslancha) [inline modules/apps/frontend-taglib/frontend-taglib-clay/src/main/java/com/liferay/frontend/taglib/clay/servlet/taglib/DropdownActionsTag.java:28]: Doing this we're overwriting any custom `cssClass` passed to the class. For example:

```
<clay:dropdown-actions
   cssClass="my-awesome-class"
/>
```

In this example `my-awesome-class` will be ignored and replaced with `component-action`

File: modules/apps/frontend-taglib/frontend-taglib-clay/src/main/resources/META-INF/resources/DropdownMenu.js [hunk 3]
```diff
@@ -49,16 +50,16 @@ export default function DropdownMenu({
...
-							<span
-								className={classNames({
-									'inline-item': label,
-									'inline-item-before': label,
-								})}
-							>
+						{icon &&
+							(label ? (
+								<span
+									className={'inline-item inline-item-before'}
```
Comment (carloslancha) [inline modules/apps/frontend-taglib/frontend-taglib-clay/src/main/resources/META-INF/resources/DropdownMenu.js:56]: We don't need to use `{'...'}` here, it's just a string.

File: modules/apps/content-dashboard/content-dashboard-web/src/main/resources/META-INF/resources/view.jsp (hunk 1 above)
Comment (dgarciasarai) [reply modules/apps/content-dashboard/content-dashboard-web/src/main/resources/META-INF/resources/view.jsp:204]: So, you mean it's better to add in directly in `ClayDropDownWithItems` component?

```js
className="dropdown-action"
```

File: modules/apps/frontend-taglib/frontend-taglib-clay/src/main/java/com/liferay/frontend/taglib/clay/servlet/taglib/DropdownActionsTag.java (hunk 2 above)
Comment (dgarciasarai) [reply modules/apps/frontend-taglib/frontend-taglib-clay/src/main/java/com/liferay/frontend/taglib/clay/servlet/taglib/DropdownActionsTag.java:28]: Ok, I'll change it 👍 Thanks!

File: modules/apps/frontend-taglib/frontend-taglib-clay/src/main/resources/META-INF/resources/DropdownMenu.js (hunk 3 above)
Comment (dgarciasarai) [reply modules/apps/frontend-taglib/frontend-taglib-clay/src/main/resources/META-INF/resources/DropdownMenu.js:56]: Oh! You're right! :) Thanks!

File: modules/apps/content-dashboard/content-dashboard-web/src/main/resources/META-INF/resources/view.jsp (hunk 1 above)
Comment (dgarciasarai) [reply modules/apps/content-dashboard/content-dashboard-web/src/main/resources/META-INF/resources/view.jsp]: I saw that the component itself it's not adding `dropdown-action` by default (only dropdown).

Comment (carloslancha) [reply modules/apps/content-dashboard/content-dashboard-web/src/main/resources/META-INF/resources/view.jsp]: > So, you mean it's better to add in directly in ClayDropDownWithItems component in DropdownMenu.js file?

Nope. `DropdownMenu.js` is the component that is being rendered when you use `clay-dropdown-menu` and `dropdown-actions`, that's why I suggested [this three options](https://github.com/liferay-frontend/liferay-portal/pull/211#issuecomment-671209332). Based on your approach I'd go for the second one.

File: modules/apps/frontend-taglib/frontend-taglib-clay/src/main/resources/META-INF/resources/DropdownMenu.js [hunk 4]
```diff
@@ -33,6 +32,7 @@ export default function DropdownMenu({
 	return (
 		<>
 			<ClayDropDownWithItems
+				className="dropdown-action"
```
Comment (carloslancha) [inline modules/apps/frontend-taglib/frontend-taglib-clay/src/main/resources/META-INF/resources/DropdownMenu.js:35]: As I commented [here](https://github.com/liferay-frontend/liferay-portal/pull/211#discussion_r468448021) this component is not always an actions dropdown

File: modules/apps/content-dashboard/content-dashboard-web/src/main/resources/META-INF/resources/view.jsp [hunk 5]
```diff
@@ -201,6 +201,7 @@ ContentDashboardAdminManagementToolbarDisplayContext contentDashboardAdminManage
 
 							<liferay-ui:search-container-column-text>
 								<clay:dropdown-actions
+									cssClass="component-action"
```
Comment (carloslancha) [inline modules/apps/content-dashboard/content-dashboard-web/src/main/resources/META-INF/resources/view.jsp:204]: Now we have here the same problem we had with `dropdownCssClass="dropdown-action"`. This is something present in every actions dropdown so we shouldn't need to specify it in every usage.

**Comment by jbalsas** [comment]
Hey @dgarciasarai, @carloslancha, just bring up that `clay:*` tags may have Server-Side rendering like in this case. They render:
- [In Java](https://github.com/liferay/liferay-portal/blob/master/modules/apps/frontend-taglib/frontend-taglib-clay/src/main/java/com/liferay/frontend/taglib/clay/servlet/taglib/DropdownActionsTag.java)
- [In JavaScript](https://github.com/liferay/liferay-portal/blob/master/modules/apps/frontend-taglib/frontend-taglib-clay/src/main/resources/META-INF/resources/DropdownMenu.js)

You need to make changes in coordination to avoid flickering, so any JS change needs to be accompanied by its corresponding Java change and viceversa.

Haven't gone through your comments in detail, so just talking in general here:
- Make sure you're actually describing real default behaviour and not specific one to decide whether to make global changes or usage changes
- Override [`processCssClasses`](https://github.com/liferay/liferay-portal/blob/master/modules/apps/frontend-taglib/frontend-taglib-clay/src/main/java/com/liferay/frontend/taglib/clay/internal/servlet/taglib/BaseContainerTag.java#L265) to add the default (or attribute-based) CSS classes of the component. The `cssClass` computed value will be applied to the root container in Java and wherever we pass it in the JS component.

> In addition I'd like to point out that, to me, the cssClass attribute should be triggerCssClass since it's being applied to the trigger, not to the dropdown. That attribute existed before but was deprecated for 7.3 in favor of cssClass, don't know the reason for this decision (probably 'cause we don't want this to be configured via taglib [...]

My reasoning for this change was based on the dual rendering model and observed previous usage of the tag. 
- If you check the Java version, it only renders a `button` (the trigger), then gets hydrated in JS mimicking that trigger.
- Most customization in this component happens in the trigger, not in the menu itself. Making access to those more straightforward and predictable made sense to me. As such, we have this hierarchy: `clay:dropdown-action` -> `clay:dropdown-menu` -> `clay:button`. This makes the API pretty obvious, since what it initially prints is a button. If we see further need for customization, we can add it.

One last note, because of the dual-rendering, you have access to the `dropdownItems` or any other prop from within the JS component, where you can make any modifications to it, like adding event handlers (primary case of `additionalProps`), custom props (like `className`), ...

**Comment by dgarciasarai** [comment]
Hey @jbalsas @carloslancha, I don't know what should I do next 🤔 With this change I'm trying to add missing CSS classes to the dropdown component (`<clay:dropdown-actions>`). I saw that:

* At the same level as CSS class `dropdown` there isn't `dropdown-action` class (as @carloslancha mention [here](https://github.com/liferay-frontend/liferay-portal/pull/211#issuecomment-669920660))
* Next to `dropdown-toggle` there isn't `component-action` (needed for the hover style)
* Span for the `label` in the `DropdownMenu` should not be rendered if there is no label (it causes misalignment between icon and its box)

I read all your feedback, I really appreciate guys 😃 but I don't know what should I change now and where. Can you help me a little bit more, please? 🙏 

Thanks!

**Comment by carloslancha** [comment]
@jbalsas Thanks.

The think is that right now that coordination between java and js is not happening in dropdown menu. I mean, since dropdown-action extends from dropdown-menu, that extends from button, in java we're rendering a button but in js the structure is different.

```
<clay:dropdown-action cssClass="my-awesome-class" />
```
Server rendered markup
```
<button class="... my-awesome-class" />
```
Client rendered markup
```
<div class="dropdown">
   <button class="... my-awesome-class" />
</div>
```

This difference wasn't a problem 'cause `dropdown` class styles makes no difference (right now) in a visual meaning. But according to Clay's dropdown actions markup, in actions dropdowns there should be a `dropdown-action` class next to `dropdown` one that, in this case, does make a visual difference.

So, I'm wondering if DropdownMenu should extend from Button since the structure we need is different...

**Comment by carloslancha** [comment]
I made some research and tests about all this stuff and based on what I found and understood:

- We can't extend ButtonTag from DropdownMenu since the root element is different for each one (`button` vs `div`), and because of it we wouldn't be ever able to keep consistency between server and client renderings.
- The fact of not extending ButtonTag implies a lot of stuff, among them:
  - Any dynamic attribute will not be applied (in java) to the trigger `button` anymore but to the root `div class="dropdown"` element, which as I said before makes sense to me, but I guess that will be a breaking change.
  - We would need to rewrite some things to make some attributes like `cssClass` to be rendered in the trigger button instead of in the root element, if that's the way we want our API to work. That seems pretty weird to me, but it's what we're doing already in the js component.

So, in the end, having in mind that in the way these tags are structured now we can't achieve real server/client rendering consistency, and that we want to apply Clay's defined markup, I identify two options:

- Apply defined markup only in the client side, which could cause flickering but is the easiest and fastest way to tackle this.
- Rewrite how DropdownMenuTag and DropdownActionsTag works (or at least the second one, but I'd say they're by hand), which implies breaking changes

I want to stress again that all this is based on my observation on how clay tags works, so maybe I don't have all the information. This is a FP2 issue so I guess we can wait until @jbalsas comes back from PTO to continue with it.

File: modules/apps/frontend-taglib/frontend-taglib-clay/src/main/resources/META-INF/resources/DropdownMenu.js [hunk 6]
```diff
@@ -49,16 +49,14 @@ export default function DropdownMenu({
...
 				trigger={
 					<ClayButton className={cssClass} {...otherProps}>
-						{icon && (
-							<span
-								className={classNames({
-									'inline-item': label,
-									'inline-item-before': label,
-								})}
-							>
+						{icon &&
```
Comment (jbalsas) [inline modules/apps/frontend-taglib/frontend-taglib-clay/src/main/resources/META-INF/resources/DropdownMenu.js:52]: Why is this change needed? If there's no `label`, we should be simply adding an empty `span`... how's that problematic?

Comment (jbalsas) [reply modules/apps/frontend-taglib/frontend-taglib-clay/src/main/resources/META-INF/resources/DropdownMenu.js:52]: I mean... I see it pushes the icon slightly down... but why? :D

Comment (jbalsas) [reply modules/apps/frontend-taglib/frontend-taglib-clay/src/main/resources/META-INF/resources/DropdownMenu.js:52]: /cc @pat270, @eduardoallegrini, @marcoscv-work, nothing urgent, we can go with this fix, but I'm curious as to why our markup is so picky here... 🤔

Comment (jbalsas) [reply modules/apps/frontend-taglib/frontend-taglib-clay/src/main/resources/META-INF/resources/DropdownMenu.js:52]: Actually, nevermind... it's just that we need `inline-item`but not `inline-item-before`, so we should simply remove tweak that to always add `inline-item`, I think.

**Comment by jbalsas** [comment]
Hey @dgarciasarai, @carloslancha, I've pushed some changes to address this. It does:
- Only add the `inline-item-before` class if we have a `label` but keep the other one. If this still causes the misalignment, then we'd need to also change how `button` renders server-side to avoid flickering buttons.
- Add the `component-action` class server-side so it renders correctly
- Add a `actionsDropdown` prop that we can use client-side to render the `*-action` classes

Could you please check both with and without JS to see what I mean and compare the different rendered markups? Moving quickly from one tab to another there should be no visual differences, signaling that both renders are in sync.

**Comment by jbalsas** [comment]
ci:test:relevant

**Comment by jbalsas** [comment]
ci:test:sf

File: modules/apps/frontend-taglib/frontend-taglib-clay/src/main/resources/META-INF/resources/DropdownMenu.js (hunk 4 above)
Comment (jbalsas) [reply modules/apps/frontend-taglib/frontend-taglib-clay/src/main/resources/META-INF/resources/DropdownMenu.js]: I added a conditional

File: modules/apps/content-dashboard/content-dashboard-web/src/main/resources/META-INF/resources/view.jsp (hunk 5 above)
Comment (jbalsas) [reply modules/apps/content-dashboard/content-dashboard-web/src/main/resources/META-INF/resources/view.jsp]: Also added it as part of the proposed solution.

**Comment by dgarciasarai** [comment]
Thank you so much @jbalsas! 😄

**Comment by jbalsas** [comment]
ok... going to go on a limb here and forward myself since the change feels trivial enough and tests are green

**Comment by jbalsas** [comment]
ci:forward

**Comment by carloslancha** [comment]
@jbalsas there's no flickering with the fix, but this doesn't feel "in sync" to me, I mean, we're not rendering the same markup in server and client, and not the same defined in Clay:

Clay:
```
<div>
   <button class="dropdown-toggle component-action ...">
      <svg ... />
   </button>
</div>
```

Server:
```
<div>
   <button class="component-action ...">
      <span class="inline-item">
         <svg ... />
      </span>
   </button>
</div>
```

Client:
```
<div class="dropdown dropdown-action">
   <button class="dropdown-toggle component-action ...">
      <span class="inline-item">
         <svg ... />
      </span>
   </button>
</div>
```

So we have three different markups. I feel it's already difficult to maintain the SSR rendering writing the html from JAVA (Why did I like Soy? 😄 ) to add more complexity by not rendering the same stuff.

**Comment by jbalsas** [comment]
> [...] this doesn't feel "in sync" to me, I mean, we're not rendering the same markup in server and client [...]

"In sync" is a perceived measurement. Having it not flicker makes it in sync, even if code is different. What matters is the end user result, not necessarily that the generated code looks exactly the same. This makes our dev APIs easier to use and reason with, which is another thing to take into consideration

> [...] and not the same defined in Clay

This is definitely true and we could change it... I think I simply updated the tag to output the existing HTML. Clay might've changed html in the process (from the glorious SOY days...), so this is definitely open to improvement. Feel free to tinker around with it and see if we can consolidate that markup 👍 

> Why did I like Soy? 😄 

No one knows 😂

**Comment by carloslancha** [comment]
> Why did I like SOY?

Why should I prefer something that allows you to have the same template for both server and client when we have the option of having to replicate the markup defined in a React component in a JAVA class using `jspWriter.write`? Double maintenance, double fun!

image

> What matters is the end user result, not necessarily that the generated code looks exactly the same.

I agree, but I think not having the same code could drive more easily to unexpected behaviors

> This makes our dev APIs easier to use and reason with, which is another thing to take into consideration

Why this makes our dev APIs easier to use? Actually in this case we're changing the way a developer would expect the tag works (as the rest of clay tags works in the same way), because if they want to add a custom css class to the dropdown root element they can't do it 'cause `cssClass` attribute refers to the trigger button, not to the root element. 🤷 

> This is definitely true and we could change it... [...] Feel free to tinker around with it and see if we can consolidate that markup.

I'll take a look 👍 


(I really hate that gif)

**Comment by jbalsas** [comment]
> [...] having to replicate the markup defined in a React component in a Java class using jspWriter.write? Double maintenance, double fun! Now in a less readable format!

100% true! Keep in mind that the Java rendering is just a bridge until we do full react rendering. I'm still not sure if the effort to do it is worth it since the markup we want to render server-side is minimal, so the Java surface is not expected to grow much more.

> I agree, but I think not having the same code could drive more easily to unexpected behaviors

True too!! 👍 

> Why this makes our dev APIs easier to use?

Because **printing the trigger is in fact the main responsibility** of the tag. You can check yourself in [the PR](https://github.com/liferay-frontend/liferay-portal/pull/2/files#diff-e7cb67f57f3dab6c3296e725698f9be1L85). The most common need in the old tag was not passing classes to the dropdown, but styling the trigger via `triggerCssClasses`. This makes that easier since it does so through a stable and consistent API rather than through implementation strings.

> [...] because if they want to add a custom css class to the dropdown root element they can't do it [...]

We still don't have a valid case for that, though, while we have the opposite need.

> (I really hate that gif)

I know 😘

**Comment by wincent** [comment]
> > (I really hate that gif)
> 
> I know 😘

solo-i-know

**Comment by jbalsas** [comment]
toolate
